gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/MultiHopsFlowToJobSpecCompiler.java

```diff
@@ -0,0 +1,389 @@
+
+          adjacentNode.setDistToSrc(newDist);
+          pq.add(adjacentNode);
+        }
+        visitedEdge.add(outGoingEdge);
+      }
+    }
+    log.error("No path found");
+    return new ArrayList<>();
+  }
+
+  /**
+   * As the base implementation here, all templates will be considered for each edge.
+   */
+  @Override
+  protected void populateEdgeTemplateMap() {
+    for (FlowEdge flowEdge:this.weightedGraph.edgeSet()) {
+      edgeTemplateMap.put(flowEdge.getEdgeIdentity(),
+          templateCatalog.get().
+              getAllTemplates().
+              stream().map(jobTemplate -> jobTemplate.getUri()).collect(Collectors.toList()));
+    }
+  }
+
+  // If path specified not existed, return false;
+  // else return true.
+  private boolean userSpecifiedPathVerificator(Map<Spec, SpecExecutor> specExecutorInstanceMap, FlowSpec flowSpec){
+    Map<Spec, SpecExecutor> tmpSpecExecutorInstanceMap = new HashMap<>();
+    List<String> userSpecfiedPath = Arrays.asList(optionalUserSpecifiedPath.get().split(","));
+    for (int i = 0 ; i < userSpecfiedPath.size() - 1 ; i ++ ) {
+      ServiceNode sourceNode = new BaseServiceNodeImpl(userSpecfiedPath.get(i));
+      ServiceNode targetNode = new BaseServiceNodeImpl(userSpecfiedPath.get(i+1));
+      if (weightedGraph.containsVertex(sourceNode) && weightedGraph.containsVertex(targetNode) &&
+          weightedGraph.containsEdge(sourceNode, targetNode)) {
+        tmpSpecExecutorInstanceMap.put(jobSpecGenerator(sourceNode, targetNode, flowSpec),
+            (((LoadBasedFlowEdgeImpl)weightedGraph.getEdge(sourceNode, targetNode)).getSpecExecutorInstance()));
+      }
+      else {
+        log.error("User Specified Path is invalid");
+        return false;
+      }
+    }
+    specExecutorInstanceMap.putAll(tmpSpecExecutorInstanceMap);
+    return true;
+  }
+
+  // Helper function for transform TopologySpecMap into a weightedDirectedGraph.
+  private void weightGraphGenerateHelper(TopologySpec topologySpec){
+    try{
+      Map<ServiceNode, ServiceNode> capabilities =
+          topologySpec.getSpecExecutor().getCapabilities().get();
+      for (Map.Entry<ServiceNode, ServiceNode> capability : capabilities.entrySet()) {
+
+        BaseServiceNodeImpl sourceNode = new BaseServiceNodeImpl(capability.getKey().getNodeName());
+        BaseServiceNodeImpl targetNode = new BaseServiceNodeImpl(capability.getValue().getNodeName());
+        // TODO: Make it generic
+        if (!weightedGraph.containsVertex(sourceNode)){
+          weightedGraph.addVertex(sourceNode);
+        }
+        if (!weightedGraph.containsVertex(targetNode)){
+          weightedGraph.addVertex(targetNode);
+        }
+
+        FlowEdge flowEdge = new LoadBasedFlowEdgeImpl
+            (sourceNode, targetNode, defaultFlowEdgeProps, topologySpec.getSpecExecutor());
+
+        // In Multi-Graph if flowEdge existed, just skip it.
+        if (!weightedGraph.containsEdge(flowEdge)) {
```

Comment:
Currently, every time `compileFlow()` is invoked the graph will be recomputed, while this is not an efficient implementation as you also mentioned before. If we implement the blacklist policy here, the blacklisted edges will never be seen in the graph during the whole execution time of GaaS which should not be the case. Whether they should be considered as part of path should be specific to certain compilation.

Reply:
Plus, when the construction of map finished it could be easier to identify if the serviceNode/Flowedge specified by user in blacklist indeed existed or not.